Use ARM templates for Key Vault tests

heaths opened the pull request:
Resolves #9059

heaths committed:
Use ARM templates for Key Vault tests

Resolves #9059

heaths commented:
/azp run net - keyvault - tests

heaths commented:
/azp run net - keyvault - tests

heaths commented:
/azp run net - keyvault - tests

heaths commented:
/azp run net - keyvault - tests

heaths commented:
/azp run net - keyvault - tests

pakrym commented on eng/pipelines/templates/jobs/archetype-sdk-tests.yml:
```diff
@@ -53,6 +54,31 @@ jobs:
+
+        # Chomp the string literal to more easily read and update the New-TestResources parameters.
+      - pwsh: >
+          $baseName = 't' + (New-Guid).ToString('n').Substring(0, 16);
```
Why do we generate the base name here and not as part of the script?

danieljurek replied:
We require a `baseName` in the script so users will use scope their resources to something recognizable, like a username. One could probably make a reasonable argument that the `baseName` generated here is the default and devs could set it directly.

pakrym replied:
Yeah, guid is not something recognizable. I think we can do a better job by default as part of the script.

heaths replied:
I don't want to generate it in the script because, as Daniel pointed out, the user needs to supply it. In fact, if anything, we could generate the name fully outside the script: whatever the user passes in is the resource name (still probably "rg-" as a prefix) and resources base name. The main point of this script is to be user-friendly. ID generation in the script with essential no input into the name is not user-friendly.

danieljurek commented on eng/pipelines/templates/jobs/archetype-sdk-tests.yml:
```diff
@@ -53,6 +54,31 @@ jobs:
+          -Force
+          -Verbose
+        displayName: "Create test resources"
+        env:
```
Azure DevOps is very good about redacting secret strings in the logs. We've been using environment variables as a safety measure to ensure that secrets are never output in the logs when that can be prevented.

weshaggard replied:
I agree with @pakrym here we shouldn't even put them in the env if we don't need to. We should just pass them as parameters. DevOps blindly hides these secrets from logs by a simple find/replace so that shouldn't be an issue.

heaths replied:
We have to. See my comments in the script. If we don't pass issecret=true variables as env vars, they aren't accessible. See https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#secret-variables. And, right now, they have to be environment variables for existing test code to use them (the step below).

weshaggard replied:
But isn't the script setting up the correct environment variables? I was sort of expecting we would pass in the secrets this script needs as variables, secrets can still be passed a parameters we do it all the time, and then the script would setup whatever variables are relevant for the tests.

pakrym reviewed (approved):
![https://media.giphy.com/media/p8GJOXwSNzQPu/giphy.gif](https://media.giphy.com/media/p8GJOXwSNzQPu/giphy.gif)

weshaggard commented on sdk/keyvault/tests.yml:
```diff
@@ -10,11 +10,13 @@ resources:
-    MaxParallel: 1
     ServiceDirectory: keyvault
     EnvVars:
-      AZURE_TENANT_ID: $(aad-azure-sdk-test-tenant-id)
```
Once these changes go through can you please file a follow-up item to remove these secrets in keyvault and clean-up the associated resources in our test sub.

heaths replied:
I opened Azure/azure-sdk-for-net#9170 to track.

weshaggard commented on eng/pipelines/templates/jobs/archetype-sdk-tests.yml:
```diff
@@ -53,6 +54,31 @@ jobs:
           packageType: sdk
           version: "$(DotNetCoreSDKVersion)"
 
+      # TODO: Remove when eng/New-TestResources.ps1 bootstraps the Az module.
```
Please add a link to the follow-up issue with this TODO.

heaths committed:
Resolve PR feedback

heaths commented:
/azp run net - keyvault - tests

heaths committed:
Re-map environment variables

Attempt to set both secret and non-secret variables at once did not work as hoped. Probably for the best, given the security ramifications.

heaths commented:
/azp run net - keyvault - tests

heaths committed:
Mask secrets while still setting env vars

heaths commented:
/azp run net - keyvault - tests
